### PR TITLE
feat: Cleanup process after execution

### PR DESCRIPTION
**Pull request by sugarnaoming**
## Context
When "sd-local" completed or interrupted, the Docker volume, processes, etc. were not deleted and were not clean.
Fixed to remove Docker volume, processes, etc.

## Objective
Added the following to the process:
- When interrupted:
  - It kills the process that was running.
  - Delete the docker volume
  - Delete the repository when using the `--src-url` option.
- When the execution is complete:
  - Delete the docker volume
  - Delete the repository when using the `--src-url` option.

## License

I confirm that this contribution is made under the terms of the license found in the root directory of this repository's source tree and that I have the authority necessary to make this contribution on behalf of its copyright owner.

### CODE REVIEW

File: launch/docker.go [hunk 1]
```diff
@@ -93,6 +101,7 @@ func (d *docker) runBuild(buildConfig buildConfig) error {
 
 	err = d.execDockerCommand(append(dockerCommandArgs, dockerCommandOptions...)...)
 	if err != nil {
+
```
Comment (yuichi10) [inline launch/docker.go:104]:
```suggestion
```

File: launch/docker_test.go [hunk 2]
```diff
@@ -216,6 +228,186 @@ func TestRunBuildWithSudo(t *testing.T) {
...
+		logrus.SetOutput(buf)
+
+		d.kill(syscall.SIGINT)
+
+		actual := buf.String()
+		expected := "failed to stop process:"
+		assert.True(t, strings.Contains(actual, expected), fmt.Sprintf("\nexpected: %s \nactual: %s\n", expected, actual))
+	})
+
+	t.Run("able to use sudo", func(t *testing.T) {
```
Comment (yuichi10) [inline launch/docker_test.go:320]: How about this?
```suggestion
	t.Run("success with sudo", func(t *testing.T) {
```

File: launch/docker_test.go [hunk 3]
```diff
@@ -216,6 +228,186 @@ func TestRunBuildWithSudo(t *testing.T) {
 	}
 }
 
+func TestDockerKill(t *testing.T) {
+	t.Run("success to kill process with no commands", func(t *testing.T) {
```
Comment (yuichi10) [inline launch/docker_test.go:232]: every test case use words of `to kill process`. It looks redundancy.
```suggestion
	t.Run("success with no commands", func(t *testing.T) {
```

File: launch/docker_test.go [hunk 4]
```diff
@@ -216,6 +228,186 @@ func TestRunBuildWithSudo(t *testing.T) {
...
+		}
+		c := newFakeExecCommand("SUCCESS_TO_KILL")
+		execCommand = c.execCmd
+		buf := bytes.NewBuffer(nil)
+		logrus.SetOutput(buf)
+		d.kill(syscall.SIGINT)
+		assert.Equal(t, "", buf.String())
+	})
+
+	t.Run("success to kill process", func(t *testing.T) {
```
Comment (yuichi10) [inline launch/docker_test.go:252]:
```suggestion
	t.Run("success", func(t *testing.T) {
```

File: launch/docker_test.go [hunk 5]
```diff
@@ -216,6 +228,186 @@ func TestRunBuildWithSudo(t *testing.T) {
...
+		buf := bytes.NewBuffer(nil)
+		logrus.SetOutput(buf)
+
+		d.kill(syscall.SIGINT)
+
+		actual := buf.String()
+		assert.Equal(t, "", actual)
+	})
+
+	t.Run("failed to kill process", func(t *testing.T) {
```
Comment (yuichi10) [inline launch/docker_test.go:286]:
```suggestion
	t.Run("failed", func(t *testing.T) {
```

File: launch/docker_test.go [hunk 6]
```diff
@@ -216,6 +228,186 @@ func TestRunBuildWithSudo(t *testing.T) {
...
+		}()
+
+		d.kill(syscall.SIGINT)
+
+		assert.Equal(t, fmt.Sprintf("sudo kill -2 %v", d.commands[0].Process.Pid), c.commands[1])
+	})
+}
+
+func TestDockerClean(t *testing.T) {
+	t.Run("success to clean", func(t *testing.T) {
```
Comment (yuichi10) [inline launch/docker_test.go:350]: every test case use words of `to clean`. It looks redundancy.

File: launch/docker_test.go [hunk 7]
```diff
@@ -216,6 +228,186 @@ func TestRunBuildWithSudo(t *testing.T) {
...
+			setupImageVersion: "latest",
+			commands:          []*exec.Cmd{},
+			useSudo:           false,
+		}
+
+		d.clean()
+		assert.Equal(t, fmt.Sprintf("docker volume rm --force %v", d.volume), c.commands[0])
+	})
+
+	t.Run("success to clean with sudo", func(t *testing.T) {
```
Comment (yuichi10) [inline launch/docker_test.go:368]:
```suggestion
	t.Run("success with sudo", func(t *testing.T) {
```

File: launch/docker_test.go [hunk 8]
```diff
@@ -216,6 +228,186 @@ func TestRunBuildWithSudo(t *testing.T) {
...
+			setupImageVersion: "latest",
+			commands:          []*exec.Cmd{},
+			useSudo:           true,
+		}
+
+		d.clean()
+		assert.Equal(t, fmt.Sprintf("sudo docker volume rm --force %v", d.volume), c.commands[0])
+	})
+
+	t.Run("failed to clean", func(t *testing.T) {
```
Comment (yuichi10) [inline launch/docker_test.go:386]:
```suggestion
	t.Run("failure", func(t *testing.T) {
```

File: launch/docker_test.go [hunk 9]
```diff
@@ -303,5 +495,25 @@ func TestHelperProcess(t *testing.T) {
 			os.Exit(1)
 		}
 		os.Exit(0)
+	case "SUCCESS_TO_KILL":
+		if subcmd == "command" {
+			time.Sleep(fakeProcessLifeTime)
+			os.Exit(0)
+		}
+		os.Exit(0)
+	case "SUCCESS_TO_KILL_DOCKER_CMD":
```
Comment (yuichi10) [inline launch/docker_test.go:504]: This process is same with "SUCCESS_TO_KILL". You can use one of them.

File: launch/docker_test.go [hunk 10]
```diff
@@ -303,5 +495,25 @@ func TestHelperProcess(t *testing.T) {
...
+		if subcmd == "sleep" {
+			time.Sleep(fakeProcessLifeTime)
+		}
+	case "FAILED_TO_KILL":
```
Comment (yuichi10) [inline launch/docker_test.go:508]: It would be better to combine it with the others.
```suggestion
	case "FAIL_TO_KILL":
```

**Comment by yuichi10** [comment]
We need to add some scm test case like below
- use sudo
- kill
- clean

File: launch/docker.go [hunk 11]
```diff
@@ -114,3 +124,88 @@ func (d *docker) execDockerCommand(args ...string) error {
...
+	err := d.execDockerCommand("volume", "rm", "--force", d.volume)
+
+	if err != nil {
+		logrus.Warn(fmt.Errorf("failed to remove volume: %v", err))
+	}
+}
+
+func (d *docker) waitForProcess(cmds []*exec.Cmd) error {
+	// Reducing this value will make the test faster.
+	// However, be sure to specify a time when you can sufficiently confirm that the process is dead.
```
Comment (yuichi10) [inline launch/docker.go:169]: Do we need this comments?
Now it looks we do not have retry test.

Comment (sugarnaoming) [reply launch/docker.go:168]: "waitaForProcess" is definitely waiting more than a second for a test.
And there are currently two or more tests where "waitForProcess" is used.
If we set this value to 500ms, I think we could reduce the test time by more than a second.

File: launch/docker_test.go (hunk 9 above)
Comment (sugarnaoming) [reply launch/docker_test.go]: thx.
`SUCCESS_TO_KILL_DOCKER_CMD` has been merged into `SUCCESS_TO_KILL`.

**Comment by sugarnaoming** [comment]
> We need to add some scm test case like below
> 
> * use sudo
> * kill
> * clean

thx.
Added test

File: launch/docker_test.go [hunk 12]
```diff
@@ -496,16 +496,12 @@ func TestHelperProcess(t *testing.T) {
...
 			time.Sleep(fakeProcessLifeTime)
 			os.Exit(0)
 		}
-		os.Exit(0)
-	case "SUCCESS_TO_KILL_DOCKER_CMD":
-		if subcmd == "sleep" {
-			time.Sleep(fakeProcessLifeTime)
-		}
-	case "FAILED_TO_KILL":
+		os.Exit(1)
```
Comment (kkisic) [inline launch/docker_test.go:503]: This `os.Exit` means `sudo kill` failure
```suggestion
		os.Exit(0)
```

File: launch/docker_test.go [hunk 13]
```diff
@@ -496,16 +496,12 @@ func TestHelperProcess(t *testing.T) {
 		}
 		os.Exit(0)
 	case "SUCCESS_TO_KILL":
-		if subcmd == "command" {
+		if subcmd == "command" || subcmd == "sleep" {
```
Comment (kkisic) [inline launch/docker_test.go:499]: How about unifying `command` and `sleep`?

File: scm/scm_test.go [hunk 14]
```diff
@@ -210,5 +326,20 @@ func TestHelperProcess(t *testing.T) {
 		os.Exit(0)
 	case "FAILED_PULL":
 		os.Exit(1)
+	case "SUCCESS_TO_KILL":
+		if subcmd == "command" {
```
Comment (kkisic) [inline scm/scm_test.go:330]: Would it be easier to understand if it was `sleep`?

File: scm/scm_test.go [hunk 15]
```diff
@@ -210,5 +326,20 @@ func TestHelperProcess(t *testing.T) {
...
+			time.Sleep(fakeProcessLifeTime)
+			os.Exit(0)
+		}
+	case "FAIL_TO_KILL":
+		if subcmd == "command" {
```
Comment (kkisic) [inline scm/scm_test.go:335]: Same as above.

File: launch/docker_test.go (hunk 12 above)
Comment (sugarnaoming) [reply launch/docker_test.go]: thx.
I fixed.

File: launch/docker_test.go (hunk 13 above)
Comment (sugarnaoming) [reply launch/docker_test.go]: thx.
I fixed.

File: scm/scm_test.go (hunk 14 above)
Comment (sugarnaoming) [reply scm/scm_test.go]: thx.
I fixed.

File: scm/scm_test.go (hunk 15 above)
Comment (sugarnaoming) [reply scm/scm_test.go]: thx.
I fixed.

File: launch/docker.go (hunk 11 above)
Comment (kkisic) [reply launch/docker.go:168]: Thx. I see.

**Pull request merged**